frontend/src/metabase/common/components/ActionButton/ActionButton.tsx

```diff
@@ -0,0 +1,152 @@
+import cx from "classnames";
+import {
+  forwardRef,
+  useCallback,
+  useEffect,
+  useImperativeHandle,
+  useRef,
+  useState,
+} from "react";
+import { match } from "ts-pattern";
+import { t } from "ttag";
+
+import CS from "metabase/css/core/index.css";
+import type { CancellablePromise } from "metabase/lib/promise";
+import { cancelable } from "metabase/lib/promise";
+import { Button, type ButtonProps, Icon } from "metabase/ui";
+
+export interface ActionButtonProps extends Omit<ButtonProps, "onClick"> {
+  // need to expose this ref to allow Tooltip to bind to the correct element
+  innerRef?: React.Ref<HTMLButtonElement>;
+  normalText?: React.ReactNode;
+  activeText?: React.ReactNode;
+  failedText?: React.ReactNode;
+  successText?: React.ReactNode;
+  useLoadingSpinner?: boolean;
+  actionFn: () => Promise<unknown>;
+  className?: string;
+  children?: React.ReactNode;
+}
+
+export type ActionButtonHandle = {
+  resetState: () => void;
+};
+
+export const ActionButton = forwardRef<ActionButtonHandle, ActionButtonProps>(
+  function ActionButtonInner(
+    {
+      normalText = t`Save`,
+      activeText = t`Saving...`,
+      failedText = t`Save failed`,
+      successText = t`Saved`,
+      useLoadingSpinner = false,
+      actionFn,
+      className = "",
+      children,
+      innerRef,
+      ...buttonProps
+    },
+    ref,
+  ) {
+    const [active, setActive] = useState(false);
+    const [result, setResult] = useState<null | "failed" | "success">(null);
+    const timeout = useRef<null | ReturnType<typeof setTimeout>>(null);
+    const actionPromise = useRef<CancellablePromise<unknown> | null>(null);
+
+    const resetTimeout = () => {
+      timeout.current && clearTimeout(timeout.current);
+    };
+
+    useEffect(() => {
+      const promise = actionPromise.current;
+      return () => {
+        resetTimeout();
+        if (promise) {
+          promise?.cancel?.();
+        }
+      };
+    }, [actionPromise]);
```

Comment:
This dependency is redundant. You still have an access to the promise when the component will be destroyed due to useRef nature.